### PR TITLE
Port debian and opensuse to preview channel

### PR DESCRIPTION
**Pull request by TravisEz13**
## PR Summary

Port debian and opensuse to preview channel

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
  - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `Dockerfile`, `.sh`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
  - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **Adds a new image or tag**
  - [ ] Not Applicable
- **OR**
  - [ ] Update [README.powershell.md](https://github.com/PowerShell/PowerShell-Docker/blob/master/assets/README.powershell.md)
  - [ ] Update [README.powershellcommunity.md](https://github.com/PowerShell/PowerShell-Docker/blob/master/assets/README.powershellcommunity.md)
  - [ ] Update [vsts-ci.yml](https://github.com/PowerShell/PowerShell-Docker/blob/master/vsts-ci.yml)

### CODE REVIEW

**Comment by TravisEz13** [comment]
@kiazhi 	Would you mind looking over these changes?  They are almost a direct port from the stable images.  Feel free to say no if you don't have time or aren't interested.

**Comment by kiazhi** [comment]
@TravisEz13 sure, i will have a look during the next few days since i'm leaving my current job and relocating soon to find a new job next year.

**Comment by TravisEz13** [comment]
cc @PaulHigin 	FYI, this might not be a great one to start with.
@kiazhi 	Thanks, great feedback.  I might merge before I get the rest, but feel free to finish your feedback.  I'm happy to finish in another PR.

File: release/preview/opensuse423/docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+# effectively removing the tar.gz and packages that are not needed from the final image.
+
+# Define arg(s) needed for the From statement
+ARG fromTag=42.3
+
+FROM opensuse:${fromTag} AS installer-env
+
+
+# Define Args for the needed to add the package
+ARG PS_VERSION=6.2.0-powershell.3
```
Comment (kiazhi) [inline release/preview/opensuse423/docker/Dockerfile]: Fix `ARG` value originated from stable release to preview

https://github.com/PowerShell/PowerShell-Docker/blob/554a54d464797db5bfc7134a9479da9ea53c23d9/release/preview/opensuse423/docker/Dockerfile#L13

```suggestion
ARG PS_VERSION=6.2.0-preview.3
```

File: release/preview/opensuse423/docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        -Command " \
+          \$ErrorActionPreference = 'Stop' ; \
+          \$ProgressPreference = 'SilentlyContinue' ; \
+          while(!(Test-Path -Path \$env:PSModuleAnalysisCachePath)) {  \
+            Write-Host "'Waiting for $env:PSModuleAnalysisCachePath'" ; \
+            Start-Sleep -Seconds 6 ; \
+          }"
+
+# Define args needed only for the labels
+ARG PS_VERSION=6.1.0
```
Comment (kiazhi) [inline release/preview/opensuse423/docker/Dockerfile]: Fix `ARG` value originated from stable release to preview to reflect in `LABEL`

https://github.com/PowerShell/PowerShell-Docker/blob/554a54d464797db5bfc7134a9479da9ea53c23d9/release/preview/opensuse423/docker/Dockerfile#L82

```suggestion
ARG PS_VERSION=6.2.0-preview.3
```

File: release/preview/opensuse423/docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+ADD ${PS_PACKAGE_URL} /tmp/linux.tar.gz
+
+# Unzip the Linux tar.gz
+RUN tar zxf /tmp/linux.tar.gz -C ${PS_INSTALL_FOLDER}
+
+# ------ Second stage ------
+# Start a new stage so we lose all the tar.gz layers from the final image
+FROM opensuse:${fromTag}
+
+# Install dependencies
```
Comment (kiazhi) [inline release/preview/opensuse423/docker/Dockerfile]: Reduce container image size from **618MB** to **550MB**
- Remove `libunwind` package from dependency https://github.com/PowerShell/PowerShell/pull/6964
- Remove unnecessary additional `RUN` that will increase container image layer
- Add `zypper clean -a` to clean up package manager cache
- Add `rm -f /var/log/zypp/history /var/log/zypper.log` to remove package manager history and log files

https://github.com/PowerShell/PowerShell-Docker/blob/554a54d464797db5bfc7134a9479da9ea53c23d9/release/preview/opensuse423/docker/Dockerfile#L40-L79

```suggestion
# Copy only the files we need from the previous stage
COPY --from=installer-env ["/opt/microsoft/powershell", "/opt/microsoft/powershell"]

# Define Args and Env needed to create links
ARG PS_INSTALL_VERSION=6
ENV PS_INSTALL_FOLDER=/opt/microsoft/powershell/$PS_INSTALL_VERSION \
    \
    # Define ENVs for Localization/Globalization
    DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=false \
    LC_ALL=en_US.UTF-8 \
    LANG=en_US.UTF-8

# Install dependencies
RUN zypper --non-interactive update --skip-interactive \
    && zypper --non-interactive install \
        glibc-locale \
        glibc-i18ndata \
        libicu \
        openssl \
        less \
    && localedef --charmap=UTF-8 --inputfile=en_US $LANG \
    \
    # Create the pwsh symbolic link that points to powershell
    && ln -s ${PS_INSTALL_FOLDER}/pwsh /usr/bin/pwsh \
    && ln -sf ${PS_INSTALL_FOLDER}/pwsh /usr/bin/pwsh-preview \
    # clean package manager cache
    && zypper clean -a \
    # remove package manager log file
    && rm -f /var/log/zypp/history /var/log/zypper.log
```

Comment (TravisEz13) [reply release/preview/opensuse423/docker/Dockerfile]: I modified this a little but I think this addressed your issue.

Perhaps we should have a size regression tests.  

I imagine a data file with known sizes (some command to regenerate this file) and the test would have a % regression that is allowed before the tests fails.

Comment (kiazhi) [reply release/preview/opensuse423/docker/Dockerfile]: All good. 👍

**Comment by TravisEz13** [comment]
I only saw the location change for the cache when reviewing with Paul.  I think that is a good idea.  Could you file an issue for that?

**Pull request merged**